### PR TITLE
PAL-390 resource-service persistence

### PR DESCRIPTION
**Pull request by dev930018**

### CODE REVIEW

**Comment by dev930018** [comment]
Resource-service persistence was made easier by working entirely with streams. This sees an update to the HadoopResourceService to support this new interface (see common)

File: hadoop-resource/src/main/java/uk/gov/gchq/palisade/service/resource/service/HadoopResourceService.java [hunk 1]
```diff
@@ -319,6 +313,7 @@ public String toString() {
                 .add("config=" + config)
                 .add("fileSystem=" + fileSystem)
                 .add("dataServices=" + dataServices)
+                .add(super.toString())
```
Comment (ac74475) [inline hadoop-resource/src/main/java/uk/gov/gchq/palisade/service/resource/service/HadoopResourceService.java:316]: doesn't need the super.toString() as this class isn't extending another class

File: hadoop-resource/src/main/java/uk/gov/gchq/palisade/service/resource/util/ResourceDetails.java [hunk 2]
```diff
@@ -0,0 +1,128 @@
...
+
+public class ResourceDetails {
+
+    public static final Pattern FILENAME_PATTERN = Pattern.compile("(?<type>.+)_(?<name>.+)\\.(?<format>.+)");
+    public static final String FORMAT_STRING = "TYPE_FILENAME.FORMAT";
+    private String fileName;
+    private String type;
+    private String format;
+
+    public ResourceDetails(final String fileName, final String type, final String format) {
```
Comment (developer01189998819991197253) [inline hadoop-resource/src/main/java/uk/gov/gchq/palisade/service/resource/util/ResourceDetails.java:36]: can you add some java doc and general descriptions as to what this file does/is needed for

File: logos/icon.svg [hunk 3]
```diff
@@ -1,7 +1,7 @@
 <?xml version="1.0" encoding="utf-8"?>
 <!-- Generator: Adobe Illustrator 16.0.0, SVG Export Plug-In . SVG Version: 6.00 Build 0)  -->
 <!DOCTYPE svg PUBLIC "-//W3C//DTD SVG 1.1//EN" "http://www.w3.org/Graphics/SVG/1.1/DTD/svg11.dtd">
-<svg version="1.1" id="Layer_1" xmlns="http://www.w3.org/2000/svg" xmlns:xlink="http://www.w3.org/1999/xlink" x="0px" y="0px"
```
Comment (developer01189998819991197253) [inline logos/icon.svg:4]: have you changed the icon?

File: logos/logo-without-txt.svg [hunk 4]
```diff
@@ -1,7 +1,7 @@
 <?xml version="1.0" encoding="utf-8"?>
 <!-- Generator: Adobe Illustrator 16.0.0, SVG Export Plug-In . SVG Version: 6.00 Build 0)  -->
 <!DOCTYPE svg PUBLIC "-//W3C//DTD SVG 1.1//EN" "http://www.w3.org/Graphics/SVG/1.1/DTD/svg11.dtd">
-<svg version="1.1" id="Layer_1" xmlns="http://www.w3.org/2000/svg" xmlns:xlink="http://www.w3.org/1999/xlink" x="0px" y="0px"
```
Comment (developer01189998819991197253) [inline logos/logo-without-txt.svg:4]: ditto?

File: logos/logo.svg [hunk 5]
```diff
@@ -1,9 +1,9 @@
 <?xml version="1.0" encoding="utf-8"?>
 <!-- Generator: Adobe Illustrator 16.0.0, SVG Export Plug-In . SVG Version: 6.00 Build 0)  -->
 <!DOCTYPE svg PUBLIC "-//W3C//DTD SVG 1.1//EN" "http://www.w3.org/Graphics/SVG/1.1/DTD/svg11.dtd">
-<svg version="1.1" id="Layer_1" xmlns="http://www.w3.org/2000/svg" xmlns:xlink="http://www.w3.org/1999/xlink" x="0px" y="0px"
-	 width="168.258px" height="197.911px" viewBox="0 0 168.258 197.911" enable-background="new 0 0 168.258 197.911"
-	 xml:space="preserve">
+<svg version="1.1" id="Layer_1" xmlns="http://www.w3.org/2000/svg" x="0px" y="0px"
```
Comment (developer01189998819991197253) [inline logos/logo.svg:4]: ditto?

File: hadoop-resource/src/main/java/uk/gov/gchq/palisade/service/resource/util/ResourceDetails.java (hunk 2 above)
Comment (dev930018) [reply hadoop-resource/src/main/java/uk/gov/gchq/palisade/service/resource/util/ResourceDetails.java]: See new comments and better class name

**Comment by dev930018** [comment]
On hold - DO NOT MERGE - until other repos are ready

File: hadoop-resource/src/main/java/uk/gov/gchq/palisade/service/resource/util/HadoopResourceDetails.java [hunk 6]
```diff
@@ -0,0 +1,142 @@
...
+    private String format;
+
+    public HadoopResourceDetails(final URI fileName, final String type, final String format) {
+        this.fileName = fileName;
+        this.type = type;
+        this.format = format;
+    }
+
+    public static HadoopResourceDetails getResourceDetailsFromFileName(final URI fileName) {
+        //get filename component
```
Comment (ac74475) [inline hadoop-resource/src/main/java/uk/gov/gchq/palisade/service/resource/util/HadoopResourceDetails.java:51]: not seen a unit test for this method

File: hadoop-resource/src/main/java/uk/gov/gchq/palisade/service/resource/service/HadoopResourceService.java [hunk 7]
```diff
@@ -153,81 +115,71 @@ private static Configuration createConfig(final Map<String, String> conf) {
...
-                                )
-                        );
-            } catch (RuntimeException e) {
-                throw e;
-            } catch (Exception e) {
-                throw new RuntimeException(e);
-            }
-        });
+    private Stream<LeafResource> getMappings(final String pathString, final Predicate<HadoopResourceDetails> predicate) {
+        //pull latest connection details
```
Comment (ac74475) [inline hadoop-resource/src/main/java/uk/gov/gchq/palisade/service/resource/service/HadoopResourceService.java:160]: Is this comment still valid?

File: hadoop-resource/src/main/java/uk/gov/gchq/palisade/service/resource/util/HadoopResourceDetails.java [hunk 8]
```diff
@@ -0,0 +1,142 @@
...
+        Matcher match = validateNameRegex(fileString);
+        if (!match.matches()) {
+            throw new IllegalArgumentException("Filename doesn't comply with " + FORMAT_STRING + ": " + fileName);
+        }
+
+        return new HadoopResourceDetails(fileName, match.group("type"), match.group("format"));
+    }
+
+    public static boolean isValidResourceName(final URI fileName) {
+        requireNonNull(fileName);
```
Comment (ac74475) [inline hadoop-resource/src/main/java/uk/gov/gchq/palisade/service/resource/util/HadoopResourceDetails.java:64]: not seen a unit test for this method, specifically interested in edge cases around the regex.

File: hadoop-resource/src/main/java/uk/gov/gchq/palisade/service/resource/service/HadoopResourceService.java [hunk 9]
```diff
@@ -153,81 +115,71 @@ private static Configuration createConfig(final Map<String, String> conf) {
...
+        final RemoteIterator<LocatedFileStatus> remoteIterator;
+        try {
+            remoteIterator = this.getFileSystem().listFiles(new Path(pathString), true);
+
+            return getPaths(remoteIterator)
```
Comment (ac74475) [inline hadoop-resource/src/main/java/uk/gov/gchq/palisade/service/resource/service/HadoopResourceService.java:165]: Want to see unit tests that validate this flow logic with different predicates. I would also like to see the connection detail random distribution being tested but not sure how that can be done with a guaranteed success. Might be worth making the distribution round robin rather than random.

File: hadoop-resource/src/main/java/uk/gov/gchq/palisade/service/resource/util/HadoopResourceDetails.java (hunk 8 above)
Comment (dev930018) [reply hadoop-resource/src/main/java/uk/gov/gchq/palisade/service/resource/util/HadoopResourceDetails.java:64]: I have no idea what that regex is testing for, could you weigh in on what Hadoop might be expecting?

Comment (ac74475) [reply hadoop-resource/src/main/java/uk/gov/gchq/palisade/service/resource/util/HadoopResourceDetails.java:64]: So all that pattern seems to be looking for is that filenames follow the pattern of <type>_<id/name>.<format> e.g. Employee_123.avro or sfsfrwe1279_sdf123.afs668. So really it isn't a brilliant pattern matcher as it is only looking for the existence of an underscore and a dot

Comment (ac74475) [reply hadoop-resource/src/main/java/uk/gov/gchq/palisade/service/resource/util/HadoopResourceDetails.java:64]: Therefore this validation is not Hadoop specific but an artefact of how we are currently auto populating the type and serialised format from the name. I would strongly recommend that once we have the resource example prepopulation, then we should remove this validation as it is no longer needed.

Comment (dev930018) [reply hadoop-resource/src/main/java/uk/gov/gchq/palisade/service/resource/util/HadoopResourceDetails.java:64]: So, it's just that sandard TYPE_FILENAME.FORMAT style, there aren't any real requirements for edge cases, but it won't allow for empty-string types, names or formats

File: hadoop-resource/src/main/java/uk/gov/gchq/palisade/service/resource/service/HadoopResourceService.java (hunk 9 above)
Comment (dev930018) [reply hadoop-resource/src/main/java/uk/gov/gchq/palisade/service/resource/service/HadoopResourceService.java:165]: We shouldn't be testing Java's rng for even distributuion really

Comment (ac74475) [reply hadoop-resource/src/main/java/uk/gov/gchq/palisade/service/resource/service/HadoopResourceService.java:164]: Hence the suggestion of changing it to do a round robin distribution rather than the current random.

File: hadoop-resource/src/test/java/uk/gov/gchq/palisade/service/resource/util/HadoopResourceDetailsTest.java [hunk 10]
```diff
@@ -0,0 +1,61 @@
...
+
+import java.io.File;
+import java.net.URI;
+
+import static org.hamcrest.MatcherAssert.assertThat;
+import static org.hamcrest.Matchers.equalTo;
+
+@RunWith(JUnit4.class)
+public class HadoopResourceDetailsTest {
+
```
Comment (ac74475) [inline hadoop-resource/src/test/java/uk/gov/gchq/palisade/service/resource/util/HadoopResourceDetailsTest.java:33]: So the intent was to test that the getResourceDetailsFromFileName copes with the different paths, e.g. /root/user/test_file.txt, ./test_file.txt as well as the filenames being incorrect. ./test.txt

File: hadoop-resource/src/main/java/uk/gov/gchq/palisade/service/resource/service/HadoopResourceService.java (hunk 9 above)
Comment (dev-958) [reply hadoop-resource/src/main/java/uk/gov/gchq/palisade/service/resource/service/HadoopResourceService.java:164]: couldn't the content of the map be moved into a private method and be unit tested as a functional?

File: hadoop-resource/src/main/java/uk/gov/gchq/palisade/service/resource/service/HadoopResourceService.java [hunk 11]
```diff
@@ -153,81 +115,72 @@ private static Configuration createConfig(final Map<String, String> conf) {
...
+                    .map(HadoopResourceDetails::getResourceDetailsFromFileName)
+                    .filter(predicate)
+                    .map(this::addConnectionDetail);
+        } catch (IOException | IllegalStateException e) {
+            LOGGER.error("Error while listing files: ", e);
+            return Stream.empty();
+        }
+    }
+
+    public LeafResource addConnectionDetail(final HadoopResourceDetails hadoopResourceDetails) {
```
Comment (ac74475) [inline hadoop-resource/src/main/java/uk/gov/gchq/palisade/service/resource/service/HadoopResourceService.java:175]: public or protected? Do you want people using this method outside of this class and the tests?

Comment (dev930018) [reply hadoop-resource/src/main/java/uk/gov/gchq/palisade/service/resource/service/HadoopResourceService.java:175]: Good spot

File: hadoop-resource/src/main/java/uk/gov/gchq/palisade/service/resource/service/HadoopResourceService.java [hunk 12]
```diff
@@ -153,81 +115,72 @@ private static Configuration createConfig(final Map<String, String> conf) {
...
-        final Predicate<ResourceDetails> predicate = detail -> request.getSerialisedFormat().equals(detail.getFormat());
-        return getFutureMappings(pathString, predicate);
+        final Predicate<HadoopResourceDetails> predicate = detail -> serialisedFormat.equals(detail.getFormat());
+        return getMappings(pathString, predicate);
     }
 
     @Override
-    public CompletableFuture<Boolean> addResource(final AddResourceRequest request) {
-        throw new UnsupportedOperationException(ERROR_ADD_RESOURCE);
+    public Boolean addResource(final LeafResource leafResource) {
```
Comment (developer01189998819991197253) [inline hadoop-resource/src/main/java/uk/gov/gchq/palisade/service/resource/service/HadoopResourceService.java:154]: add javadoc along the lines of: user should not be able to add resources manually, therefore throw error
